### PR TITLE
[SPARK-35581][SQL] Support special datetime values in typed literals only

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to support special datetime values introduced by #25708 and by #25716 only in typed literals, and don't recognize them in parsing strings to dates/timestamps. The following string values are supported only in typed timestamp literals:
- `epoch [zoneId]` - `1970-01-01 00:00:00+00 (Unix system time zero)`
- `today [zoneId]` - midnight today.
- `yesterday [zoneId]` - midnight yesterday
- `tomorrow [zoneId]` - midnight tomorrow
- `now` - current query start time.

For example:
```sql
spark-sql> SELECT timestamp 'tomorrow';
2019-09-07 00:00:00
```

Similarly, the following special date values are supported only in typed date literals:
- `epoch [zoneId]` - `1970-01-01`
- `today [zoneId]` - the current date in the time zone specified by `spark.sql.session.timeZone`.
- `yesterday [zoneId]` - the current date -1
- `tomorrow [zoneId]` - the current date + 1
- `now` - the date of running the current query. It has the same notion as `today`.

For example:
```sql
spark-sql> SELECT date 'tomorrow' - date 'yesterday';
2
``` 

### Why are the changes needed?
In the current implementation, Spark supports the special date/timestamp value in any input strings casted to dates/timestamps that leads to the following problems:
- If executors have different system time, the result is inconsistent, and random. Column values depend on where the conversions were performed.
- The special values play the role of distributed non-deterministic functions though users might think of the values as constants. 

### Does this PR introduce _any_ user-facing change?
Yes but the probability should be small.

### How was this patch tested?
By running existing test suites:
```
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z interval.sql"
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z date.sql"
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z timestamp.sql"
$ build/sbt "test:testOnly *DateTimeUtilsSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon FYI, I did't add new tests because special values in typed literals are tested in `date.sql`/`timestamp.sql`.

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -91,6 +91,8 @@ license: |
 
   - In Spark 3.2, `CREATE TABLE AS SELECT` with non-empty `LOCATION` will throw `AnalysisException`. To restore the behavior before Spark 3.2, you can set `spark.sql.legacy.allowNonEmptyLocationInCTAS` to `true`.
 
+  - In Spark 3.2, the special datetime values such as `epoch`, `today`, `yesterday`, `tomorrow` and `now` are supported in typed literals only, for instance `select timestamp'now'`. In Spark 3.1 and earlier, such special values are supported in any casts of strings to dates/timestamps. To restore the behavior before Spark 3.2, you should preprocess string columns and convert the strings to desired timestamps explicitly using UDF for instance.
```
Comment (yaooqinn) [inline docs/sql-migration-guide.md:94]: In Spark 3.2, ~the~ special datetime values..... in typed literals only, for instance **(add',')** `select timestamp'now'`. In Spark 3.1 and ~earlier~ (3.0?)

Comment (MaxGekk) [reply docs/sql-migration-guide.md:94]: @yaooqinn What do you mean by: for instance **(add',')** `select timestamp'now'`. I didn't get the problem. BTW, you could use the suggestion feature, so, I would just commit your suggestions.

File: docs/sql-migration-guide.md [hunk 2]
```diff
@@ -91,6 +91,8 @@ license: |
 
   - In Spark 3.2, `CREATE TABLE AS SELECT` with non-empty `LOCATION` will throw `AnalysisException`. To restore the behavior before Spark 3.2, you can set `spark.sql.legacy.allowNonEmptyLocationInCTAS` to `true`.
 
+  - In Spark 3.2, special datetime values such as `epoch`, `today`, `yesterday`, `tomorrow` and `now` are supported in typed literals only, for instance `select timestamp'now'`. In Spark 3.1 and 3.0, such special values are supported in any casts of strings to dates/timestamps. To restore the behavior before Spark 3.2, you should preprocess string columns and convert the strings to desired dates/timestamps explicitly using UDF for instance.
```
Comment (yaooqinn) [inline docs/sql-migration-guide.md:94]:
```suggestion
  - In Spark 3.2, special datetime values such as `epoch`, `today`, `yesterday`, `tomorrow`, and `now` are supported in typed literals only, for instance, `select timestamp'now'`. In Spark 3.1 and 3.0, such special values are supported in any casts of strings to dates/timestamps. To restore the behavior before Spark 3.2, you should preprocess string columns and convert the strings to desired dates/timestamps explicitly using UDF for instance.
```

File: docs/sql-migration-guide.md (hunk 1 above)
Comment (yaooqinn) [reply docs/sql-migration-guide.md]: Hi @MaxGekk, thanks for your suggestion. 

I think if users need to preprocess the data, we may not call it as`To restore the behavior before Spark 3.2`

Comment (MaxGekk) [reply docs/sql-migration-guide.md]: What do you propose? How about `To have the behavior before Spark 3.2 ...`

File: docs/sql-migration-guide.md (hunk 2 above)
Comment (MaxGekk) [reply docs/sql-migration-guide.md:94]: hmm, I am not sure all those commas are mandatory, especially before `and`.

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon Are you ok with the changes in general?

File: docs/sql-migration-guide.md (hunk 1 above)
Comment (yaooqinn) [reply docs/sql-migration-guide.md]: How about "to keep these special values as datetimes in Spark 3.1 and 3.0, you need to match them manually, e.g. `if(c in ('now', 'today'), current_date() else c)`.

I think it's better to suggest user use builtin functions than UDFs

**Review by cloud-fan — APPROVED** [review]

File: docs/sql-migration-guide.md (hunk 1 above)
Comment (MaxGekk) [reply docs/sql-migration-guide.md]: I agree that builtin functions is better suggestion. Let's me update this.

**Review by yaooqinn — APPROVED** [review]

File: docs/sql-migration-guide.md [hunk 3]
```diff
@@ -91,6 +91,8 @@ license: |
 
   - In Spark 3.2, `CREATE TABLE AS SELECT` with non-empty `LOCATION` will throw `AnalysisException`. To restore the behavior before Spark 3.2, you can set `spark.sql.legacy.allowNonEmptyLocationInCTAS` to `true`.
 
+  - In Spark 3.2, special datetime values such as `epoch`, `today`, `yesterday`, `tomorrow`, and `now` are supported in typed literals only, for instance, `select timestamp'now'`. In Spark 3.1 and 3.0, such special values are supported in any casts of strings to dates/timestamps. To keep these special values as dates/timestamps in Spark 3.1 and 3.0, you should replace them manually, e.g. `if (c in ('now', 'today'), current_date(), cast(c as date))`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md:94]: > if (c in ('now', 'today'), current_date(), cast(c as date))

What does it mean?

Comment (MaxGekk) [reply docs/sql-migration-guide.md:94]: @cloud-fan See the example:
```scala
scala> val df = Seq("now", "2021-01-19", "today").toDF("c")
df: org.apache.spark.sql.DataFrame = [c: string]

scala> df.selectExpr("if (c in ('now', 'today'), current_date(), cast(c as date))").show(false)
+----------------------------------------------------------+
|(IF((c IN (now, today)), current_date(), CAST(c AS DATE)))|
+----------------------------------------------------------+
|2021-06-01                                                |
|2021-01-19                                                |
|2021-06-01                                                |
+----------------------------------------------------------+
```

**Comment by MaxGekk** [comment]
GA passed. Merging to master.
Thank you @yaooqinn and @cloud-fan for your reviews.

**Comment by gengliangwang** [comment]
@MaxGekk QQ: why do the special values support zone id? I tried PostgreSQL and this is not supported.
Also, the zone id in the special string is not respected and won't affect the result.  However, for a normal timestamp string the zone id is respected:
```
scala> spark.sql("select cast(timestamp'1970-01-01 00:00:00 UTC' as long)").show()
+-----------------------------------------------+
|CAST(TIMESTAMP '1970-01-01 08:00:00' AS BIGINT)|
+-----------------------------------------------+
|                                              0|
+-----------------------------------------------+


scala> spark.sql("select cast(timestamp'1970-01-01 00:00:00 CET' as long)").show()
+-----------------------------------------------+
|CAST(TIMESTAMP '1970-01-01 07:00:00' AS BIGINT)|
+-----------------------------------------------+
|                                          -3600|
+-----------------------------------------------+


```

**Comment by gengliangwang** [comment]
I suggest that we remove the support of zone id in the special strings to make things simple.

**Comment by MaxGekk** [comment]
> QQ: why do the special values support zone id? I tried PostgreSQL and this is not supported.

@gengliangwang We follow PostgreSQL behavior strictly. Time zones are accepted but ignored, see the tests from PostgreSQL:
https://github.com/postgres/postgres/blob/master/src/test/regress/sql/timestamp.sql#L19-L20

**Comment by gengliangwang** [comment]
@MaxGekk Oh, I made a mistake in the test with PostgreSQL. Sorry for that.

**Pull request closed without merging**